handlers/authentication.go

```diff
@@ -696,14 +719,125 @@ func OpenIdRedirect(w http.ResponseWriter, r *http.Request) {
 		url.QueryEscape(fmt.Sprintf("%x-%s", csrfHash, nowTime.UTC().Format("060102150405"))),
 	)
 	http.Redirect(w, r, redirectUri, http.StatusFound)
 }
+
+func OpenIdCodeFlowHandler(w http.ResponseWriter, r *http.Request) bool {
+	// Read received HTTP params and check for data completeness
+	openIdParams, err := business.ExtractOpenIdCallbackParams(w, r)
+	if err != nil {
+		RespondWithError(w, http.StatusBadRequest, err.Error())
+		return true // Return true to mark request as handled (because an error is already being sent back)
+	}
+
+	if checkFailure := business.CheckOpenIdAuthenticationCodeFlowParams(openIdParams); len(checkFailure) != 0 {
+		log.Infof("Not handling OpenId code flow authentication: %s", checkFailure)
+		return false
+	}
+
+	// Start handling the callback for OpenId authentication (code flow)
+
+	// CSRF mitigation
+	if stateError := business.ValidateOpenIdState(openIdParams); len(stateError) > 0 {
+		RespondWithError(w, http.StatusForbidden, fmt.Sprintf("Request rejected: %s", stateError))
+		return true
+	}
+
+	// Exchange the received code for a token
+	// err := business.RequestOpenIdToken(code, httputil.GuessKialiURL(r))
+	if err := business.RequestOpenIdToken(openIdParams, httputil.GuessKialiURL(r)); err != nil {
+		RespondWithDetailedError(w, http.StatusForbidden, "failure when retrieving user identity", err.Error())
+		return true
+	}
+
+	if err := business.ParseOpenIdToken(openIdParams); err != nil {
+		RespondWithError(w, http.StatusUnauthorized, err.Error())
+		return true
+	}
+
+	// Replay attack mitigation
+	if nonceError := business.ValidateOpenIdNonceCode(openIdParams); len(nonceError) > 0 {
+		RespondWithError(w, http.StatusForbidden, fmt.Sprintf("OpenId token rejected: %s", nonceError))
+		return true
+	}
+
+	// Check if user trying to login has enough privileges to login
+	httpStatus, errMsg, detailedError := business.VerifyOpenIdUserAccess(openIdParams.IdToken)
+	if detailedError != nil {
+		RespondWithDetailedError(w, httpStatus, errMsg, detailedError.Error())
+		return true
+	} else if httpStatus != http.StatusOK {
+		RespondWithError(w, httpStatus, errMsg)
+		return true
+	}
+
+	// Create Kiali's session cookie and set it in the response
+
+	// For the OpenId's "authentication code" flow we don't want
+	// any of the session data to be readable even in the browser's
+	// developer console. So, we cipher the session data using AES-GCM
+	// which allows to leave aside the usage of JWT tokens. So, this
+	// builds a bare JSON serialized into a string, cipher it and
+	// set a cookie with the ciphered string. Yet, we use the
+	// "IanaClaims" type just for convenience to avoid creating new types and
+	// to bring some type convergence on types for the auth source code.
+	sessionData := business.BuildOpenIdJwtClaims(openIdParams)
+	sessionDataJson, err := json.Marshal(sessionData)
+	if err != nil {
+		RespondWithDetailedError(w, http.StatusInternalServerError, "Error when creating credentials", err.Error())
```

Comment:
I see the same error message repeated here and 3 other places. To make it easier on someone debugging this, let's make each error message different so when someone says, `I see "Error when creating credentials" in the logs` we won't have to wonder which of the 4 places it happened in the code.

```suggestion
		RespondWithDetailedError(w, http.StatusInternalServerError, "Error when creating credentials - failed to marshal json", err.Error())
```

Reply:
+1000 unique messages are priceless.  If you want the same text add a numeric or some sort of identifier.

Reply:
This is using `RespondWithDetailedError`. I was hoping that the "detailed" part would give context. But yes, it doesn't hurt if we add the context ourselves.

Reply:
The detailed part does give context - but sometimes it doesn't, and because it's runtime generated, we can't just grep for it to find where it came from.

When a user hits an error, what do they say? "I have an error in my logs that says "Error when creating credentials" with details "foo bar"". We then (at least I know I do) immediately do a grep for "Error when creating credentials" to find out where it came from in the code. In this case, we'll find 4 places and we don't know which one to immediately look at... "foo bar" may or may not help make it more obvious, but even then we aren't absolutely sure and we have to look at the 4 places to eliminate the possibilities. And of course we can't grep on "foo bar" because that isn't in the code.

For this reason, I never (or I try to never) have the same exact error message in multiple places. In fact, I've been known to do little tricks like add a period at the end of one sentence in one message, but leave it off the sentence in the other :) Anything to make it unique so someone debugging can quickly find the one spot where the log/error message came from by just doing a simple grep.